### PR TITLE
feat(MBS-14442): implement Guess Punctuation button with single quote conversion logic

### PR DESCRIPTION
**Pull request by arsinclair**
- my first PR here, so please let me know if there's something I missed;
- this is narrowly scoped to single quotes detection and replacement only; once I'm confirmed this is the direction we want to move in, I'll extend the logic to double quotes, dashes, ellipses, etc in future PRs.

# Problem

[MBS-14442](https://tickets.metabrainz.org/browse/MBS-14442).

Add a **Guess Punctuation** button similar to the existing **Guess Case** functionality.

When clicked, the button should automatically detect ASCII punctuation in entered titles and replace it with the appropriate preferred Unicode punctuation where applicable.

This issue is intentionally limited to implementing the punctuation-guessing/replacement functionality. Highlighting changes, character-level diffs, undo support, and improvements to Guess Case should be handled separately.

This is a smaller scoped part of [MBS-11646](https://tickets.metabrainz.org/browse/MBS-11646).

# Solution

The conversion logic, as well as the collection of test cases is based on (or entirely copied from) @kellnerd's [Guess Unicode punctuation](https://github.com/kellnerd/musicbrainz-scripts/blob/main/README.md#guess-unicode-punctuation) script, which is amazing and I've been using it for years.

---

Most of the logic is just a replica of Guess Case functionality with a different handler function just for punctuation.

One oddity to mention, in https://github.com/metabrainz/musicbrainz-server/pull/3833/changes#diff-f013e29e5dd3ed65921c393fa6ec6b329e2929a36fd87b67c2a4cc602f180557R449 the spacing between buttons is achieved with a literal white space character and its size is not constrained to integer pixels. So whereas 9px space allocation worked for 3 buttons, for 4 buttons setting it to 12px is not enough, since the width of a single white space is 5.28px.

image

So I had to set it to 13px.

```js
    input.with-guesscase-guessfeat {
        width: @form-input-width - (@form-icon-size * 4) - 13 !important;
    }
```

Screenshots:

image
image
image

# AI usage

Yes, plumbing the new logic based on the existing Guess Case functionality, adding tests and full review.

# Testing

All automated tests were ran by the LLM.

Manual testing and polishing done with a local musicbrainz docker container.

# Documenting

:beginner: If you updated documentation pages, mention them here, such as:
Updated [WikiDocs page](https://wiki.musicbrainz.org/Category:WikiDocs_Page)

Not yet, I need a confirmation first if this is the right direction.

### CODE REVIEW

File: root/static/scripts/guess-punctuation.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * @returns {string} - The string with guessed punctuation.
+ */
+export default function guessPunctuation(inputString: string): string {
+  return inputString
+    // 'n' is an elision rather than text enclosed in single quotes.
+    .replace(/(^|\W)'(n)'(?=\W|$)/gi, '$1’$2’')
+    // Match quote pairs only when they are bounded by non-word characters.
+    .replace(/(^|[^\p{L}\d])'(.+?)'(?=[^\p{L}\d]|$)/gu, '$1‘$2’')
+    // Any remaining single quotes are apostrophes.
+    .replace(/'/g, '’');
```
Comment (kellnerd) [inline root/static/scripts/guess-punctuation.js]: I would actually advice against using this as the last step in the official MBS implementation. It also converts weird edge cases which should better be left alone.
This was OK for a userscript implementation, although I wanted to improve this for some time now, but never got to it.

File: root/static/scripts/tests/GuessPunctuation.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ * This file is part of MusicBrainz, the open internet music database,
+ * and is licensed under the GPL version 2, or (at your option) any
+ * later version: http://www.gnu.org/licenses/gpl-2.0.txt
+ */
+
+import test from 'tape';
+
+import guessPunctuation from '../guess-punctuation.js';
+
+test('Guess punctuation', function (t) {
```
Comment (kellnerd) [inline root/static/scripts/tests/GuessPunctuation.js:18]: I know [these examples](https://github.com/kellnerd/es-utils/blob/main/string/punctuation.test.ts) 😁
Not a licensing expert (do they also apply for test data?) and I don't intend to sue MetaBrainz, but maybe add a comment here to satisfy the MIT license?

File: root/static/scripts/guess-punctuation.js (hunk 1 above)
Comment (arsinclair) [reply root/static/scripts/guess-punctuation.js]: This is the only logic that handles single apostrophes, if we remove it we won't be able to handle them. If we change the order, it will affect paired apostrophes.

Of course, there's a always a way of going forward with the whitelist, e.g. `'s[\s^]`, `n't[\s^]` etc, but I don't know if this is necessary at this point. I suppose it would be helpful to see some examples of those edge cases if you still have them?

File: root/static/scripts/tests/GuessPunctuation.js (hunk 2 above)
Comment (arsinclair) [reply root/static/scripts/tests/GuessPunctuation.js:18]: Yes, I took'em directly from your script, there's no need to reinvent the wheel when you can build on the shoulders of giants 😁 

I assumed that this is just a compilation of titles from existing MB data, so I am not sure that that licensing applies here, although I am also not a licensing expert either.

To be honest, adding a MIT block just within this file feels quite odd to me for mainly two reasons:
- the test cases will continue to evolve, and the new additions will have to fall under the GPL license, which MBS repo has.
- when someone bundles and uses this software, they look at the top level License file (hopefully they do) but not necessarily going through all files and checking whether the license is different from the top-level one, so this interleaving of licenses is just a bit of a surprise for someone who might want to stay compliant.

I'm happy to give credit where credit is due and I should have done it right in the PR description. Fixed that now.
Whether to add the MIT block in the tests file or not, I'd like to see some more opinions shared. It doesn't cost me a penny and I'm totally happy to do it, just that it feels a bit odd.
An alternative is to just search the MB and pick up similar, but different examples.

Comment (mwiencek) [reply root/static/scripts/tests/GuessPunctuation.js:18]: We have some precedent for that here: https://github.com/metabrainz/musicbrainz-server/blob/748e737/root/static/scripts/release-editor/utils.js#L8-L11

Edit: I think crediting kellnerd for the curated set of example cases with a link to the MIT license should suffice. (It's not the individual titles themselves that are his creative work.)

Comment (arsinclair) [reply root/static/scripts/tests/GuessPunctuation.js:18]: Very good, added that now.

**Comment by arsinclair** [comment]
Hey @mwiencek, this has been sitting here for more than a week. Does the review normally take that long, or is there any changes needed in the PR?

**Comment by reosarevok** [comment]
We're a small team and MusicBrainz specifically is even more time constrained than usual right now, so things can take a bit, sorry :)

**Comment by arsinclair** [comment]
No worries, I will work on my patience 😁

**Comment by reosarevok** [comment]
If it works, can you send a pull request to us for that too? It sounds useful :face_with_peeking_eye:

**Comment by arsinclair** [comment]
> If it works, can you send a pull request to us for that too? It sounds useful 🫣

A pull request for what? This is the only pull request in question, I think.

**Comment by reosarevok** [comment]
For patience I meant! :joy:

**Comment by arsinclair** [comment]
LOL, it took me a while to get it hahahaha

Once I have figured it out, I'll be sure to open a PR for the patience

**Comment by arsinclair** [comment]
I think I'll just focus on userscripts as it allows for much faster feedback loop.

Please let me know when you are accepting contributions again and if I have time I will resubmit the PR.

**Comment by mwiencek** [comment]
@arsinclair Sorry for the delay. I'd like to keep this PR open if you don't mind. I can push changes if you're not able to work on it. Reading through the above discussion, I'm interested in hearing @kellnerd's response to [your comment above](https://github.com/metabrainz/musicbrainz-server/pull/3833#discussion_r3780458445), but I know he's been quite busy too.

**Comment by arsinclair** [comment]
OK, I asked Github to restore my fork, which thankfully they just did.

It was a little odd to see this sitting in the quiet for some time, considering that the change is trivial in my eyes and that other bigger PRs are merged. (I knew about the bot storm problem, sorry about that and that it kept you busy)

I didn't realise you were waiting for kellnerd's to check in, but ultimately if he has no time for this, we need to find another way forward. If you would like me to maybe sift through the MB data and try to find some edge cases, I could do that but I initially didn't want to go that deep, considering that the change is non-destructive and even if there are some edge cases, it is easy to roll back.

---

I have 4 or 5 more PRs regarding this feature if this one is succesfully merged, so I calculated that with this speed this will drag for the whole year hence I closed it 😅

**Comment by reosarevok** [comment]
Even if something takes the whole year, it is still better than not implementing it at all :) (that said, userscripts are of course faster, and you can always implement the next steps you want as a userscript first and test it that way / get some community feedback)

**Comment by arsinclair** [comment]
For this one, it's based on the Guess Unicode script by kellnerd, I guess it's already well tested.

File: root/static/scripts/guess-punctuation.js (hunk 1 above)
Comment (mwiencek) [reply root/static/scripts/guess-punctuation.js]: A couple examples where this breaks:

 * 4'33"
 * Blue Hawai'i

(I'd guess there are several other kinds of quote-like characters which shouldn't be substituted by an apostrophe.) So I get the feeling a whitelist is indeed safer for an MBS implementation, even if it's more to maintain.

Comment (arsinclair) [reply root/static/scripts/guess-punctuation.js]: I implemented a white-list approach and just pushed a commit, but now I am wondering whether a white-list solution is a good idea after all. Yes, we can do a decent job for English, but there's many other languages where this would require separate whitelists. E.g. French `l'amour` should be converted, but it won't be according to English rules. The problem is that there's an unknown number of possible combinations.

Comment (reosarevok) [reply root/static/scripts/guess-punctuation.js]: It's better to convert only some cases than to over-convert IMO (because keeping the ASCII symbols is not as wrong as wrongly using the Unicode ones). So a whitelist makes sense to me - more entries can be added to it when requested. [Ll]'[A-z] might be a reasonable thing to put in the whitelist from the beginning, since French does use those a lot? Or even play it safer and do something like [Ll]'[AaEeHhIiOoUu] I guess..

Comment (arsinclair) [reply root/static/scripts/guess-punctuation.js]: Okay, then let's review it as is. I just added the French case
